examples/caustics.dx

```diff
@@ -0,0 +1,533 @@
+    False -> case y0 < y1 of
+      True -> (y0, y1, x0, x1, dy, dx)
+      False -> (y1, y0, x1, x0, dy, dx)
+
+  slope = (v1 - v0)/(u1 - u0)
+
+  min_u_int = FToI $ floor (u0 / du)
+  max_u_int = FToI $ ceil (u1 / du)
+  -- TODO: once support for dynamic lists and also plotting lands, replace this
+  -- with iteration over a dynamic list, something like
+  --    u_values = arange min_u_int (max_u_int + 1)
+  --    for i.
+  --      u_int = u_values.i
+  --      ...
+  delta_u = max_u_int - min_u_int + 1
+  for u_offset:(Fin delta_u).
+    u_int = ordinal u_offset + min_u_int
+
+    -- Convert u back into relative coordinates
+    u_rel = IToF u_int * dx
+    -- Use that to figure out indices for v
+    v_rel = v0 + (u_rel - u0) * slope
+    v_int_prev = FToI $ floor (v_rel / dv)
+    v_int_next = case v1 > v0 of
+      True -> v_int_prev + 1
+      False -> v_int_prev - 1
+    
+    -- Rasterize into the two boxes (u_int, v_int_prev), (u_int, v_int_next)
+    case x_changes_more of
+      True ->
+        rasterLineSegmentInFourPixels u_int v_int_prev a b fill canvas
+        rasterLineSegmentInFourPixels u_int v_int_next a b fill canvas
+      False ->
+        rasterLineSegmentInFourPixels v_int_prev u_int a b fill canvas
+        rasterLineSegmentInFourPixels v_int_next u_int a b fill canvas
+  ()
+
+
+:plotmat
+  s = 100
+  drawIntoCanvas s s $ \canvas.
+    rasterLineSegment (0.1, 0.3) (0.8, 0.6) 1.0 canvas
+    rasterLineSegment (0.6, 0.8) (0.3, 0.1) 1.0 canvas
+
+-- Bilinear approximation still preserves line integral, confirming we are
+-- rendering the right set of pixels.
+sparseRenderedLineIntegral = sum $ sum $ drawIntoCanvas 100 100 $
+  rasterLineSegment (0.1, 0.3) (0.8, 0.6) 1.0
+
+:p trueLineIntegral ~~ sparseRenderedLineIntegral
+
+'## Light path visualization
+We can visualize light by rasterizing many lines coming from a single point.
+
+def invertAndGammaCorrect (raw : w=>h=>Float) : w=>h=>Float =
+  -for i j. exp $ 0.45 * log raw.i.j
+
+:plotmat
+  s = 100
+  n = 10000
+  invertAndGammaCorrect $ drawIntoCanvas s s $ \canvas.
+    for i:(Fin n).
+      key = ixkey (newKey 42) i
+      angle = 2. * pi * rand key
+      rasterLineSegment (0.5, 0.5) (0.5 + (cos angle), 0.5 + (sin angle)) 1.0 canvas
+    ()
+
+'But it's not very interesting unless the light has something to interact with!
+
+data BoundaryCurve =
+  LineSegment start:RelPoint end:RelPoint
+  ArcSegment center:RelPoint radius:Float theta1:Float theta2:Float  -- smallest theta must be between [-pi, pi)
+
+Ray = { start: RelPoint & dir: RelPoint }
+
+-- TODO: rewrite with table unpacking
+def invert2d (m: Fin 2 => Fin 2 => Float) : Fin 2 => Fin 2 => Float =
+  det = m.(0@_).(0@_) * m.(1@_).(1@_) - m.(0@_).(1@_) * m.(1@_).(0@_)
+  (1.0/det) .* [ [m.(1@_).(1@_), -m.(0@_).(1@_)]
+  , [-m.(1@_).(0@_), m.(0@_).(0@_)]
+  ]
+
+-- TODO: deduplicate with https://github.com/google-research/dex-lang/pull/276
+def atan2 (y:Float) (x:Float) :Float =
```

Comment:
Whoops, I forgot you already added an `atan2` function.  The one I used is more expensive, but follows the TF and XLA implementations.  I think its derivation is pretty close to the one you used, but with a higher-degree polynomial.  I don't know if the XLA people had a good reason for their choice - perhaps it should depend on the floating-point precision being used.

Reply:
"Do what XLA does" is probably a better justification for an implementation than "use the first thing you see on stack overflow" anyways!